File: ruby/lib/trailer_vote/media_types/partials/image_links.rb
# frozen_string_literal: true

require 'media_types'

module TrailerVote
  module MediaTypes
    module Partials
      IMAGE_LINKS = ::MediaTypes::Scheme.new do
        link :self

        link :original do
          attribute :content_digest, String, optional: true
          attribute :width, Numeric, optional: true
          attribute :height, Numeric, optional: true
        end

        %i[thumbnail xlarge large medium small xsmall].each do |size|
          link size, optional: true do
            attribute :content_digest, String, optional: true
            attribute :width, Numeric, optional: true
            attribute :height, Numeric, optional: true
          end
        end
      end
    end
  end
end

File: typescript/application/partials/image_links.ts
import { ProductImageTypeV1 } from '../types/product_image_type'

import { ElementType, literals } from '../utils/literals'

export const ImageSizes = literals(
  'thumbnail',
  'xlarge',
  'large',
  'medium',
  'small',
  'xsmall'
)

export type ImageSize = ElementType<typeof ImageSizes>

export type ImageLink = {
  href: string
  content_digest?: string
  width?: number
  height?: number
}

export type ImageLinks = {
  _links: {
    self: {
      href: string
      type?: ProductImageTypeV1
    }

    original?: ImageLink
    thumbnail?: ImageLink
    xlarge?: ImageLink
    large?: ImageLink
    medium?: ImageLink
    small?: ImageLink
    xsmall?: ImageLink

    /*
    [K in ImageSize]: {
      href: string
      content_digest?: string
      width?: number
      height?: number
    }
    */
  }
}
